ReactNativeClient/lib/services/searchengine/queryBuilder.ts

```diff
@@ -0,0 +1,361 @@
+	let result = new Array(tags.length).fill(`${intersect ? 'INTERSECT' : 'UNION'} SELECT note_tags.note_id FROM note_tags WHERE note_tags.tag_id IN (SELECT tags.id from tags WHERE tags.title LIKE ?)`).join(' ');
+	result = intersect ? `SELECT note_tags.note_id as id FROM note_tags WHERE 1 ${result}` : `SELECT note_tags.note_id as id FROM note_tags WHERE 0 ${result}`;
+	result = negated ? `tag_filter_negated as (${result})` : `tag_filter as (${result})`;
+	return result;
+};
+
+
+// WITH RECURSIVE resource_filter as (
+// 	SELECT note_resources.note_id as id FROM note_resources WHERE 1 INTERSECT SELECT note_resources.note_id FROM note_resources WHERE note_resources.is_associated=1
+// 	AND note_resources.resource_id IN (SELECT resources.id from resources WHERE resources.mime LIKE 'application/pdf'))
+
+const resourceFilter = (mimeTypes: string[], intersect: boolean = true, negated: boolean = false) => {
+	let result = new Array(mimeTypes.length).fill(`${intersect ? 'INTERSECT' : 'UNION'} SELECT note_resources.note_id FROM note_resources WHERE note_resources.is_associated=1 AND note_resources.resource_id IN (SELECT resources.id from resources WHERE resources.mime LIKE ?)`).join(' ');
+	result = intersect ? `SELECT note_resources.note_id as id FROM note_resources WHERE 1 ${result}` : `SELECT note_resources.note_id as id FROM note_resources WHERE 0 ${result}`;
+	result = negated ? `resource_filter_negated as (${result})` : `resource_filter as (${result})`;
+	return result;
+};
+
+const negatedAnyResourceFilter = (mimeTypes: string[]) => {
+	let result = new Array(mimeTypes.length).fill('select * from (select * from notes_with_resources except SELECT note_resources.note_id FROM note_resources WHERE note_resources.is_associated=1 AND note_resources.resource_id IN (select resources.id from resources where resources.mime like ?))').join(' UNION ');
+	result = `resource_filter_negated(id) as (${result})`;
+	return result;
+};
+
+// WITH RECURSIVE tagged_notes
+// AS
+// (
+// SELECT DISTINCT note_tags.note_id FROM note_tags
+// ),
+// tag_filter_negated(id)
+// as
+// (
+// select * from (
+// select * from tagged_notes except SELECT note_tags.note_id FROM note_tags WHERE note_tags.tag_id IN (select tags.id from tags where tags.title like 'tag3')
+// )
+// UNION
+// select * from (
+// select * from tagged_notes except SELECT note_tags.note_id FROM note_tags WHERE note_tags.tag_id IN (select tags.id from tags where tags.title like 'tag1')
+// )
+// )
+
+
+const negatedAnyTagFilter = (tags: string[]) => {
+	let result = new Array(tags.length).fill('select * from (select * from tagged_notes except SELECT note_tags.note_id FROM note_tags WHERE note_tags.tag_id IN (select tags.id from tags where tags.title like ?))').join(' UNION ');
+	result = `tag_filter_negated(id) as (${result})`;
+	return result;
+};
+
+
+const noteBookFilter = (names: string[], negated: boolean = false) => {
```

Comment:
Another thing is why are there optional parameters on all these functions? I didn't check all of them but for example `noteBookFilter` is called only once with only the `names` argument. So the `negated` argument is always false. Any unnecessary argument like this should be removed, which will also help make the code easier to understand.